things/service/src/main/java/org/eclipse/ditto/things/service/persistence/actors/strategies/commands/MigrateThingDefinitionStrategy.java

```diff
@@ -0,0 +1,279 @@
+                            final Result<ThingEvent<?>> mergeResult = mergeStrategy.doApply(
+                                    context,
+                                    existingThing,
+                                    nextRevision,
+                                    mergeThingCommand,
+                                    metadata
+                            );
+
+                            return CompletableFuture.completedFuture(mergeResult);
+                        })
+        );
+    }
+
+    private JsonObject evaluatePatchConditions(final Thing existingThing,
+            final JsonObject migrationPayload,
+            final Map<ResourceKey, String> patchConditions,
+            final DittoHeaders dittoHeaders) {
+        final JsonObjectBuilder adjustedPayloadBuilder = migrationPayload.toBuilder();
+
+        for (Map.Entry<ResourceKey, String> entry : patchConditions.entrySet()) {
+            final ResourceKey resourceKey = entry.getKey();
+            final String conditionExpression = entry.getValue();
+
+            final boolean conditionMatches = evaluateCondition(existingThing, conditionExpression, dittoHeaders);
+
+            final JsonPointer resourcePointer = JsonFactory.newPointer(resourceKey.getResourcePath());
+            if (!conditionMatches && doesMigrationPayloadContainResourceKey(migrationPayload, resourcePointer)) {
+                adjustedPayloadBuilder.remove(resourcePointer);
+            }
+        }
+
+        return adjustedPayloadBuilder.build();
+    }
+    public boolean doesMigrationPayloadContainResourceKey(JsonObject migrationPayload, JsonPointer pointer) {
+        return migrationPayload.getValue(pointer).isPresent();
+    }
+    private boolean evaluateCondition(final Thing existingThing,
+            final String conditionExpression,
+            final DittoHeaders dittoHeaders) {
+        try {
+            final var criteria = QueryFilterCriteriaFactory
+                    .modelBased(RqlPredicateParser.getInstance())
+                    .filterCriteria(conditionExpression, dittoHeaders);
+
+            final var predicate = ThingPredicateVisitor.apply(criteria,
+                    PlaceholderFactory.newPlaceholderResolver(TIME_PLACEHOLDER, new Object()));
+
+            return predicate.test(existingThing);
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    private CompletionStage<Thing> generateSkeleton(final Context<ThingId> context,
+            final MigrateThingDefinition command,
+            final DittoHeaders dittoHeaders) {
+        return wotThingSkeletonGenerator.provideThingSkeletonForCreation(
+                        command.getEntityId(),
+                        ThingsModelFactory.newDefinition(command.getThingDefinitionUrl()),
+                        dittoHeaders
+                )
+                .thenApply(optionalSkeleton -> optionalSkeleton.orElseThrow());
```

Comment:
`orElseThrow` is not good here .. it will throw an `IllegalArgumentException` which will not contain any useful information about why this failed.
I would suggest to throw a more meaningful subclass of `WotException`, also maintaining `dittoHeaders` to have proper propagation of correlation-id, trace, etc.